# Segmenting and Clustering Neighborhoods in Toronto

#### Import libraries

In [1]:
from bs4 import BeautifulSoup
import requests
import pandas as pd
import numpy as np
from sklearn.cluster import KMeans
# !conda install -c conda-forge folium=0.5.0 --yes
import folium
import matplotlib.cm as cm
import matplotlib.colors as colors
import math

## 1) Data Wrangling

#### Import HTML code from Wikipedia and create BeautifulSoup object

In [2]:
source = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

soup = BeautifulSoup(source, 'lxml')

#### Extract information from the table and put it into a list

In [3]:
# Find the right table on the wikipedia site 
table = soup.find('table')

# Write table entries into a list
data = []
for entry in table.find_all('td'):
    data.append(entry.get_text(strip=True))

#### Bring the list into the right shape

In [4]:
# Reshape list and convert to pandas dataframe
data_r = np.reshape(data, (288,3))
df = pd.DataFrame(data_r, columns = ['PostalCode', 'Borough', 'Neighborhood'])
df.head()

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

#### Delete rows where borough is not assigned

In [5]:
df = df.drop(df[df['Borough'] == 'Not assigned'].index)
df.reset_index(inplace = True ,drop=True)
df.head()

PostalCode           Borough      Neighborhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M5A  Downtown Toronto       Regent Park
4        M6A        North York  Lawrence Heights

#### Group Neighborhoods with equal PostalCode

In [6]:
df = df.groupby(['PostalCode', 'Borough'])['Neighborhood'].apply(", ".join).reset_index()
df.head()

PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Rouge, Malvern
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

#### Find "not assigned" neighborhood(s) in the dataframe

In [7]:
na = df.loc[df['Neighborhood'].str.match('Not assigned')] 
print(na)

   PostalCode       Borough  Neighborhood
85        M7A  Queen's Park  Not assigned


#### Change found neighborhood(s) to associated borough(s)

In [8]:
for i in na.index:
    df['Neighborhood'][i] = df['Borough'][i]

#### Shape of the final table

In [9]:
df.shape

(103, 3)

#### Final table

In [10]:
df.head()

PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Rouge, Malvern
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

## 2) Neighborhood Exploration

#### Import coordinates of the neighborhoods

In [11]:
import types
from botocore.client import Config
import ibm_boto3

def __iter__(self): return 0

@hidden_cell
# The following code accesses a file in your IBM Cloud Object Storage. It includes your credentials.
# You might want to remove those credentials before you share your notebook.
client_cc63f8d7f7d240dfb8542250416d6165 = ibm_boto3.client(service_name='s3',
    ibm_api_key_id='e7Dhl4U5i-PjYLuSeWXPxiEuhywuDuNfGh_b2_VN4Hdw',
    ibm_auth_endpoint="https://iam.bluemix.net/oidc/token",
    config=Config(signature_version='oauth'),
    endpoint_url='https://s3-api.us-geo.objectstorage.service.networklayer.com')

body = client_cc63f8d7f7d240dfb8542250416d6165.get_object(Bucket='capstonesegmentation-donotdelete-pr-hqzlt8uazakfly',Key='Geospatial_Coordinates.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

coord_df = pd.read_csv(body)
coord_df.columns = ['PostalCode', 'Latitude', 'Longitude']
coord_df.head()



PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

#### Add the coordinates to the neighborhood table

In [12]:
df2 = pd.merge(df, coord_df , on='PostalCode')
df2.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [13]:
# Coordinates of Toronto
latitude = 43.653908
longitude = -79.384293
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653908, -79.384293.


#### Show the neighborhoods on a map

In [14]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df2['Latitude'], df2['Longitude'], df2['Borough'], df2['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [15]:
@hidden_cell
CLIENT_ID = 'LUBUJEE50HNRFRVKFQS0VPEMNCQT2MF3FXRHXT0RY5FSXK2B' # your Foursquare ID
CLIENT_SECRET = 'QVXUMQCNCKY55HAWFREPC2EWNZENSMJ0XWNQYH131KDYEHHN' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

# print('Your credentails:')
# print('CLIENT_ID: ' + CLIENT_ID)
# print('CLIENT_SECRET:' + CLIENT_SECRET)

#### Define a function that gets venues within a radius of 500m from a neighborhood (with Foursquare)

In [16]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Retrieve $($maximally $100)$ venues for each neighborhood

In [17]:
LIMIT = 100
toronto_venues = getNearbyVenues(names=df2['Neighborhood'],
                                   latitudes=df2['Latitude'],
                                   longitudes=df2['Longitude']
                                  )

Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
Silver Hills, York Mills
Newtonbrook, Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park, Don Mills South
Bathurst Manor, Downsview North, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens, Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West, 

In [18]:
print(toronto_venues.shape)
toronto_venues.head()

(2265, 7)


Neighborhood  Neighborhood Latitude  \
0                          Rouge, Malvern              43.806686   
1  Highland Creek, Rouge Hill, Port Union              43.784535   
2  Highland Creek, Rouge Hill, Port Union              43.784535   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                            Venue  Venue Latitude  \
0              -79.194353                          Wendy's       43.807448   
1              -79.160497                RIGHT WAY TO GOLF       43.785177   
2              -79.160497            Royal Canadian Legion       43.782533   
3              -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   
4              -79.188711                G & G Electronics       43.765309   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.161108           Golf Course  
2       -79.163085                   Bar  
3       -79.189914           Pizza Place  
4       -79.191537     Electronics Store

#### Number of different neighborhoods in the table

In [19]:
len(set(toronto_venues['Neighborhood']))

99

The amount of neighborhoods before was $103$. This means that for 4 neighborhoods no venues were found.

#### Number of venues found in each neighborhood

In [20]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Agincourt                                                               4   
Agincourt North, L'Amoreaux East, Milliken, Ste...                      2   
Albion Gardens, Beaumond Heights, Humbergate, J...                     11   
Alderwood, Long Branch                                                  8   
Bathurst Manor, Downsview North, Wilson Heights                        21   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      22   
Berczy Park                                                            58   
Birch Cliff, Cliffside West                                             4   
Bloordale Gardens, Eringate, Markland Wood, Old...                      8   
Brockton, Exhibition Place, Parkdale Village                           21   
Business Reply Mail Processing Centre 969 Eastern                      19   
CFB Toronto, Downsview East                                             2   
CN Tower, Bathurst Quay, Island airport, Harbou...                     17   
Cabbagetown, St. James Town                                            43   
Caledonia-Fairbanks                                                     6   
Canada Post Gateway Processing Centre                                  11   
Cedarbrae                                                               7   
Central Bay Street                                                     85   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               16   
Church and Wellesley                                                   86   
Clairlea, Golden Mile, Oakridge                                         9   
Clarks Corners, Sullivan, Tam O'Shanter                                11   
Cliffcrest, Cliffside, Scarborough Village West                         3   
Cloverdale, Islington, Martin Grove, Princess G...                      2   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        8   
...                                                                   ...   
Northwest                                                               2   
Northwood Park, York University                                         7   
Parkdale, Roncesvalles                                                 15   
Parkwoods                                                               3   
Queen's Park                                                           41   
Rosedale                                                                5   
Roselawn                                                                2   
Rouge, Malvern                                                          1   
Runnymede, Swansea                                                     39   
Ryerson, Garden District                                              100   
Scarborough Village                                                     3   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        98   
Studio District                                                        41   
The Annex, North Midtown, Yorkville                                    24   
The Beaches                                                             5   
The Beaches West, India Bazaar                                         19   
The Danforth West, Riverdale                                           41   
The Junction North, Runnymede                                           4   
The Kingsway, Montgomery Road, Old Mill North   

#### Amount of different venue types in the Toronto area

In [21]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 276 uniques categories.


#### Use one hot encoding with the venues category variable

In [22]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Trail  Train Station  \
0                 0                    0  ...      0              0   
1                 0                    0  ...      0              0   
2                 0                    0  ...      0              0   
3                 0                    0  ...      0              0   
4                 0                    0  ...      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 276 columns]

#### Find the percentage of each venue category in each neighborhood

In [23]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide, King, Richmond     0.000000   
1                                           Agincourt     0.000000   
2   Agincourt North, L'Amoreaux East, Milliken, St...     0.000000   
3   Albion Gardens, Beaumond Heights, Humbergate, ...     0.000000   
4                              Alderwood, Long Branch     0.000000   
5     Bathurst Manor, Downsview North, Wilson Heights     0.000000   
6                                     Bayview Village     0.000000   
7                   Bedford Park, Lawrence Manor East     0.000000   
8                                         Berczy Park     0.000000   
9                         Birch Cliff, Cliffside West     0.000000   
10  Bloordale Gardens, Eringate, Markland Wood, Ol...     0.000000   
11       Brockton, Exhibition Place, Parkdale Village     0.047619   
12  Business Reply Mail Processing Centre 969 Eastern     0.000000   
13                        CFB Toronto, Downsview East     0.000000   
14  CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   
15                        Cabbagetown, St. James Town     0.000000   
16                                Caledonia-Fairbanks     0.000000   
17              Canada Post Gateway Processing Centre     0.000000   
18                                          Cedarbrae     0.000000   
19                                 Central Bay Street     0.011765   
20          Chinatown, Grange Park, Kensington Market     0.000000   
21                                           Christie     0.000000   
22                               Church and Wellesley     0.011628   
23                    Clairlea, Golden Mile, Oakridge     0.000000   
24            Clarks Corners, Sullivan, Tam O'Shanter     0.000000   
25    Cliffcrest, Cliffside, Scarborough Village West     0.000000   
26  Cloverdale, Islington, Martin Grove, Princess ...     0.000000   
27                     Commerce Court, Victoria Hotel     0.000000   
28                                         Davisville     0.000000   
29                                   Davisville North     0.000000   
..                                                ...          ...   
69                                          Northwest     0.000000   
70                    Northwood Park, York University     0.000000   
71                             Parkdale, Roncesvalles     0.000000   
72                                          Parkwoods     0.000000   
73                                       Queen's Park     0.024390   
74                                           Rosedale     0.000000   
75                                           Roselawn     0.000000   
76                                     Rouge, Malvern     0.000000   
77                                 Runnymede, Swansea     0.000000   
78                           Ryerson, Garden District     0.000000   
79                                Scarborough Village     0.000000   
80                                     St. James Town     0.000000   
81                    Stn A PO Boxes 25 The Esplanade     0.000000   
82                                    Studio District     0.024390   
83                The Annex, North Midtown, Yorkville     0.000000   
84                                        The Beaches     0.000000   
85                     The Beaches West, India Bazaar     0.000000   
86                       The Danforth West, Riverdale     0.024390   
87                      The Junction North, Runnymede     0.000000   
88      The Kingsway, Montgomery Road, Old Mill North     0.000000   
89                                   Thorncliffe Park     0.066667   
90                                   Victoria Village     0.000000   
91                                          Westmount     0.000000   
92                                             Weston     0.000000   
93                                   Willowdale South     0.000000   
94                                    Willowdale We

#### Give the $5$ most common venue types for each neighborhood

In [24]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.07
1             Café  0.05
2              Bar  0.04
3       Steakhouse  0.04
4  Thai Restaurant  0.04


----Agincourt----
            venue  freq
0          Lounge  0.25
1  Breakfast Spot  0.25
2    Skating Rink  0.25
3  Clothing Store  0.25
4     Yoga Studio  0.00


----Agincourt North, L'Amoreaux East, Milliken, Steeles East----
                       venue  freq
0                       Park   0.5
1                 Playground   0.5
2                Yoga Studio   0.0
3  Middle Eastern Restaurant   0.0
4                      Motel   0.0


----Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown----
                 venue  freq
0        Grocery Store  0.18
1          Pizza Place  0.18
2          Coffee Shop  0.09
3       Sandwich Place  0.09
4  Fried Chicken Joint  0.09


----Alderwood, Long Branch----
            venue  freq
0     Pizza Place  0.25
1

#### Function that sorts the most common venues in descending order

In [25]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

#### Show a table with the $10$ most common venues for each neighborhood

In [26]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                          Agincourt        Breakfast Spot   
2  Agincourt North, L'Amoreaux East, Milliken, St...                  Park   
3  Albion Gardens, Beaumond Heights, Humbergate, ...         Grocery Store   
4                             Alderwood, Long Branch           Pizza Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café            Steakhouse       Thai Restaurant   
1                Lounge          Skating Rink        Clothing Store   
2            Playground            Donut Shop    Dim Sum Restaurant   
3           Pizza Place           Coffee Shop              Pharmacy   
4           Coffee Shop                   Gym          Skating Rink   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Bar   American Restaurant      Asian Restaurant   
1             Drugstore        Discount Store              Dive Bar   
2                 Diner        Discount Store              Dive Bar   
3        Sandwich Place            Beer Store   Fried Chicken Joint   
4              Pharmacy                   Pub        Sandwich Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                 Hotel      Sushi Restaurant                    Gym  
1               Dog Run      Doner Restaurant             Donut Shop  
2               Dog Run      Doner Restaurant              Drugstore  
3  Fast Food Restaurant           Video Store    Dumpling Restaurant  
4              Dive Bar          Dessert Shop     Dim Sum Restaurant

## 3) Clustering of the neighborhoods

#### Train K-means algorithm on the grouped neighborhood set (one-hot encoded + averaged)

In [27]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 1, 4, 4, 0, 0, 0, 0, 0], dtype=int32)

#### Add cluster labels to the table

In [28]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df2

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             3.0  Fast Food Restaurant             Drugstore   
1 -79.160497             0.0                   Bar           Golf Course   
2 -79.188711             4.0     Electronics Store    Mexican Restaurant   
3 -79.216917             0.0           Coffee Shop     Korean Restaurant   
4 -79.239476             0.0      Hakka Restaurant   Fried Chicken Joint   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                 Diner        Discount Store              Dive Bar   
1         Women's Store            Donut Shop                 Diner   
2   Rental Car Location          Intersection           Pizza Place   
3      Insurance Office             Drugstore                 Diner   
4    Athletics & Sports                Bakery                  Bank   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0               Dog Run      Doner Restaurant            Donut Shop   
1        Discount Store              Dive Bar               Dog Run   
2        Breakfast Spot        Medical Center         Women's Store   
3        Discount Store              Dive Bar               Dog Run   
4  Caribbean Restaurant       Thai Restaurant         Women's Store   

  9th Most Common Venue 10th Most Common Venue  
0   Dumpling Restaurant     College Rec Center  
1      Doner Restaurant              Drugstore  
2        Discount Store               Dive Bar  
3      Doner Restaurant             Donut Shop  
4      Doner Restaurant               Dive Bar

#### Show different clusters on the map

The neighborhoods in which no venues were found are treated as outliers and are marked in grey on the map.

In [29]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    if math.isnan(cluster) == True:
        folium.CircleMarker(
            [lat, lon],
            radius=5,
            popup=label,
            color='grey',
            fill=True,
            fill_color='grey',
            fill_opacity=0.7).add_to(map_clusters)
    else:
        folium.CircleMarker(
            [lat, lon],
            radius=5,
            popup=label,
            color=rainbow[int(cluster)-1],
            fill=True,
            fill_color=rainbow[int(cluster)-1],
            fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Show the most common venues for each of the clusters and neighborhoods

In [30]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels          1st Most Common Venue  \
1        Scarborough             0.0                            Bar   
3        Scarborough             0.0                    Coffee Shop   
4        Scarborough             0.0               Hakka Restaurant   
6        Scarborough             0.0                 Discount Store   
7        Scarborough             0.0                         Bakery   
8        Scarborough             0.0                  Movie Theater   
9        Scarborough             0.0                College Stadium   
10       Scarborough             0.0              Indian Restaurant   
11       Scarborough             0.0                    Auto Garage   
12       Scarborough             0.0                 Breakfast Spot   
15       Scarborough             0.0             Chinese Restaurant   
17        North York             0.0                           Pool   
18        North York             0.0                 Clothing Store   
19        North York             0.0                           Café   
22        North York             0.0                    Coffee Shop   
26        North York             0.0               Basketball Court   
27        North York             0.0                            Gym   
28        North York             0.0                    Coffee Shop   
29        North York             0.0             Falafel Restaurant   
32        North York             0.0                     Food Truck   
33        North York             0.0             Athletics & Sports   
34        North York             0.0          Portuguese Restaurant   
36         East York             0.0                   Skating Rink   
37      East Toronto             0.0           Other Great Outdoors   
38         East York             0.0                    Coffee Shop   
39         East York             0.0              Indian Restaurant   
41      East Toronto             0.0               Greek Restaurant   
42      East Toronto             0.0                            Gym   
43      East Toronto             0.0                           Café   
45   Central Toronto             0.0                           Park   
..               ...             ...                            ...   
61  Downtown Toronto             0.0                    Coffee Shop   
62        North York             0.0             Italian Restaurant   
63   Central Toronto             0.0                         Garden   
64   Central Toronto             0.0                           Park   
65   Central Toronto             0.0                    Coffee Shop   
66  Downtown Toronto             0.0                           Café   
67  Downtown Toronto             0.0                           Café   
68  Downtown Toronto             0.0                Airport Service   
69  Downtown Toronto             0.0                    Coffee Shop   
70  Downtown Toronto             0.0                    Coffee Shop   
71        North York             0.0         Furniture / Home Store   
72        North York             0.0               Asian Restaurant   
73              York             0.0                          Field   
75  Downtown Toronto             0.0                  Grocery Store   
76      West Toronto             0.0                    Supermarket   
77      West Toronto             0.0                            Bar   
78      West Toronto             0.0                    Coffee Shop   
79        North York             0.0                           Park   
80              York             0.0                     Restaurant   
82      West Toronto             0.0                           Café   
83      West Toronto             0.0                 Breakfast Spot   
84      West Toronto             0.0                           Café   
85      Queen's Park             0.0                    Coffee Shop   
86       Mississauga             0.0                          Hotel   
87      East Toronto             0.0 

In [31]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14       Scarborough             1.0                  Park   
23        North York             1.0                  Park   
30        North York             1.0                  Park   
31        North York             1.0                  Park   
44   Central Toronto             1.0                  Park   
50  Downtown Toronto             1.0                  Park   
74              York             1.0                  Park   
90         Etobicoke             1.0                  Park   
98              York             1.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14            Playground            Donut Shop    Dim Sum Restaurant   
23                  Bank     Convenience Store             Drugstore   
30               Airport             Drugstore                 Diner   
31         Grocery Store                  Bank         Shopping Mall   
44              Bus Line           Swim School    Falafel Restaurant   
50            Playground              Building                 Trail   
74                Market  Fast Food Restaurant         Women's Store   
90                  Pool                 River     Electronics Store   
98     Convenience Store           Event Space  Ethiopian Restaurant   

          5th Most Common Venue        6th Most Common Venue  \
14                        Diner               Discount Store   
23               Discount Store                     Dive Bar   
30               Discount Store                     Dive Bar   
31                   Donut Shop                        Diner   
44                        Diner               Discount Store   
50            Electronics Store  Eastern European Restaurant   
74                     Pharmacy         Colombian Restaurant   
90  Eastern European Restaurant          Dumpling Restaurant   
98          Empanada Restaurant            Electronics Store   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
14                     Dive Bar               Dog Run      Doner Restaurant   
23                      Dog Run      Doner Restaurant            Donut Shop   
30                      Dog Run      Doner Restaurant            Donut Shop   
31               Discount Store              Dive Bar               Dog Run   
44                     Dive Bar               Dog Run      Doner Restaurant   
50          Dumpling Restaurant   Empanada Restaurant             Drugstore   
74      Comfort Food Restaurant  Ethiopian Restaurant       College Stadium   
90                    Drugstore            Donut Shop      Doner Restaurant   
98  Eastern European Restaurant   Dumpling Restaurant          Dessert Shop   

   10th Most Common Venue  
14              Drugstore  
23    Dumpling Restaurant  
30    Dumpling Restaurant  
31       Doner Restaurant  
44             Donut Shop  
50       Department Store  
74    Empanada Restaurant  
90          Deli / Bodega  
98              Drugstore

In [32]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
102  Etobicoke             2.0             Drugstore   Rental Car Location   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
102            Donut Shop                 Diner        Discount Store   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
102              Dive Bar               Dog Run      Doner Restaurant   

    9th Most Common Venue 10th Most Common Venue  
102         Women's Store           Dessert Shop

In [33]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0   Scarborough             3.0  Fast Food Restaurant             Drugstore   
25   North York             3.0  Fast Food Restaurant     Food & Drink Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Diner        Discount Store              Dive Bar   
25                  Park  Gym / Fitness Center      Department Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                Dog Run      Doner Restaurant            Donut Shop   
25  Ethiopian Restaurant   Empanada Restaurant     Electronics Store   

          9th Most Common Venue 10th Most Common Venue  
0           Dumpling Restaurant     College Rec Center  
25  Eastern European Restaurant    Dumpling Restaurant

In [34]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
2    Scarborough             4.0     Electronics Store    Mexican Restaurant   
5    Scarborough             4.0           Pizza Place            Playground   
13   Scarborough             4.0           Pizza Place   Fried Chicken Joint   
24    North York             4.0         Grocery Store              Pharmacy   
35     East York             4.0  Fast Food Restaurant           Pizza Place   
40     East York             4.0                  Park           Pizza Place   
81          York             4.0               Brewery              Bus Line   
89     Etobicoke             4.0           Pizza Place           Coffee Shop   
95     Etobicoke             4.0     Convenience Store                  Park   
96    North York             4.0           Pizza Place   Empanada Restaurant   
99     Etobicoke             4.0           Pizza Place    Chinese Restaurant   
100    Etobicoke             4.0                  Park              Bus Line   
101    Etobicoke             4.0         Grocery Store           Pizza Place   

    3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
2     Rental Car Location          Intersection                Pizza Place   
5       Convenience Store         Women's Store                 Donut Shop   
13        Thai Restaurant    Italian Restaurant              Shopping Mall   
24            Pizza Place           Coffee Shop             Discount Store   
35                   Café                  Bank         Athletics & Sports   
40            Coffee Shop     Convenience Store              Women's Store   
81            Pizza Place         Grocery Store                  Drugstore   
89                    Gym          Skating Rink                   Pharmacy   
95                   Café          Liquor Store                   Pharmacy   
96          Women's Store            Donut Shop                      Diner   
99            Coffee Shop        Sandwich Place  Middle Eastern Restaurant   
100           Pizza Place     Mobile Phone Shop                    Dog Run   
101           Coffee Shop              Pharmacy             Sandwich Place   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2          Breakfast Spot        Medical Center         Women's Store   
5      Dim Sum Restaurant                 Diner        Discount Store   
13                   Bank          Noodle House              Pharmacy   
24                Butcher         Women's Store               Dog Run   
35              Gastropub          Intersection              Pharmacy   
40             Donut Shop                 Diner        Discount Store   
81         Discount Store              Dive Bar               Dog Run   
89                    Pub        Sandwich Place              Dive Bar   
95            Pizza Place           Coffee Shop            Beer Store   
96         Discount Store              Dive Bar               Dog Run   
99          Women's Store                 Diner        Discount Store   
100    Dim Sum Restaurant                 Diner        Discount Store   
101            Beer Store   Fried Chicken Joint  Fast Food Restaurant   

    9th Most Common Venue 10th Most Common Venue  
2          Discount Store               Dive Bar  
5                Dive Bar                Dog Run  
13   Fast Food Restaurant     Chinese Restaurant  
24           Dessert Shop     Dim Sum Restaurant  
35         Breakfast Spot   Gym / Fitness Center  
40               Dive Bar                Dog Run  
81       Doner Restaurant             Donut Shop  
89           Dessert Shop     Dim Sum Restaurant  
95             Donut Shop         Discount Store  
96       Doner Restaurant    Dumpling Restaurant  
99               Dive Bar                Dog Run  
100              Dive Bar       Doner Restaurant  
101           Video Store    Dumpling Restaurant